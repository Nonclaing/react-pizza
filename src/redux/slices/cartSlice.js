import { createSlice } from '@reduxjs/toolkit';

export const cartSlice = createSlice({
  name: 'cart',
  initialState: {
    totalPrice: 0,
    items: {},
    count: 0,
  },
  reducers: {
    setItemCount(state, action) {
      const oldItemCount = state.items[action.payload.item.id].count || 0;
      const newCount = oldItemCount - action.payload.count;
      state.items[action.payload.item] = { ...action.payload.item, count: action.payload.count };
      state.count -= newCount;
      state.totalPrice -= newCount * state.items[action.payload.item.id].price;
    },
    addItem(state, action) {
      if (state.items[action.payload.item.id]) {
        state.items[action.payload.item.id].count += 1;
      } else {
        state.items[action.payload.item.id] = { ...action.payload.item, count: 1 };
      }

      state.count += 1;
      state.totalPrice += state.items[action.payload.item.id].price;
    },
    removeItem(state, action) {
      if (!state.items[action.payload.item.id]) return;

      state.items[action.payload.item.id].count -= 1;

      state.count -= 1;
      state.totalPrice -= state.items[action.payload.item.id].price;

      if (state.items[action.payload.item.id].count <= 0)
        delete state.items[action.payload.item.id];
    },
    clearItem(state, action) {
      if (!state.items[action.payload.item.id]) return;

      state.count -= state.items[action.payload.item.id].count;
      state.totalPrice -=
        state.items[action.payload.item.id].price * state.items[action.payload.item.id].count;

      delete state.items[action.payload.item.id];
    },
    clearItems(state, action) {
      state.count = 0;
      state.items = {};
      state.totalPrice = 0;
    },
  },
});

export const selectCart = (state) => state.cart;

export const { setItemCount, addItem, removeItem, clearItems, clearItem } = cartSlice.actions;

export default cartSlice.reducer;
