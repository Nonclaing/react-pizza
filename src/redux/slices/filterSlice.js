import { createSlice } from '@reduxjs/toolkit';
import { list as sortList } from './../../components/Sort/Sort';

export const filterSlice = createSlice({
  name: 'filter',
  initialState: {
    categoryId: 0,
    currentPage: 1,
    sort: {
      id: 0,
      text: 'популярности (воз.)',
      sortBy: 'rating',
      sortOrder: 'ask',
    },
  },
  reducers: {
    setCategoryId(state, action) {
      state.categoryId = action.payload;
    },
    setSearchValue(state, action) {
      state.searchValue = action.payload;
    },
    setSortValue(state, action) {
      state.sort = action.payload;
    },
    setCurrentPageCount(state, action) {
      state.currentPage = action.payload;
    },
    setFilters(state, action) {
      const currentPage = Number(action.payload.currentPage);
      const sortId = Number(action.payload.sortId);
      const categoryId = Number(action.payload.categoryId);
      const sort = sortList.find(({ id }) => id === sortId);

      state.currentPage = currentPage ? currentPage : state.currentPage;
      state.sort = sort ? sort : state.sort;
      state.categoryId = categoryId ? categoryId : state.categoryId;
    },
  },
});
export const selectFilter = (state) => state.filter;
export const selectFilterSort = (state) => state.filter.sort;
export const selectCategoryId = (state) => state.filter.categoryId;

export const { setCategoryId, setSortValue, setCurrentPageCount, setFilters } = filterSlice.actions;

export default filterSlice.reducer;
