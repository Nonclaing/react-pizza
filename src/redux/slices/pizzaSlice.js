import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';
import qs from 'qs';

export const fetchPizzas = createAsyncThunk(
  'pizza/fetch',
  async ({ categoryId, sort, searchValue, currentPage, limit = 4 }, thunkAPI) => {
    const urlParamsString = qs.stringify({
      category: categoryId === 0 ? undefined : categoryId,
      sortBy: sort?.sortBy,
      order: sort?.sortOrder,
      search: searchValue,
      page: currentPage,
      limit,
    });

    const { data } = await axios.get(
      `https://65de266ddccfcd562f5665ae.mockapi.io/items?${urlParamsString}`,
    );
    return data;
  },
);

export const getPizza = createAsyncThunk('pizza/get', async ({ id }, thunkAPI) => {
  const { data } = await axios.get(`https://65de266ddccfcd562f5665ae.mockapi.io/items/${id}`);
  return data;
});

export const pizzaSlice = createSlice({
  name: 'pizza',
  initialState: {
    pizzaItems: [],
    pizzaDetail: {},
    loading: true,
    rejected: false,
  },
  reducers: {
    setItems(state, action) {
      state.pizzaItems = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchPizzas.pending, (state) => {
      state.loading = true;
      state.rejected = false;
    });
    builder.addCase(fetchPizzas.fulfilled, (state, action) => {
      state.pizzaItems = action.payload;
      state.loading = false;
    });
    builder.addCase(fetchPizzas.rejected, (state, action) => {
      state.loading = false;
      state.rejected = true;
    });
    builder.addCase(getPizza.pending, (state) => {
      state.loading = true;
      state.rejected = false;
    });
    builder.addCase(getPizza.fulfilled, (state, action) => {
      state.pizzaDetail = action.payload;
      state.loading = false;
    });
    builder.addCase(getPizza.rejected, (state, action) => {
      state.loading = false;
      state.rejected = true;
    });
  },
});

export const selectPizza = (state) => state.pizza;

export const selectPizzaDetail = (state) => state.pizza.pizzaDetail;

export const { setItems } = pizzaSlice.actions;

export default pizzaSlice.reducer;
