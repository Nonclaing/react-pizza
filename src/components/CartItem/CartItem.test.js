import CartItem from './CartItem';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import * as reduxHooks from 'react-redux';
import * as action from '../../redux/slices/cartSlice';
import { addItem, removeItem } from '../../redux/slices/cartSlice';

jest.mock('react-redux');

const mockDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('Тест компонента CartItem', () => {
  it('Рендер', () => {
    useSelector.mockReturnValue({
      count: 1,
      items: [
        {
          id: '0',
          type: 0,
          sizes: 26,
          title: 'Пепперони Фреш с перцем',
          price: 803,
          imageUrl:
            'https://media.dodostatic.net/image/r:584x584/11EE7D6130241E75B0AB33725248C0D0.avif',
        },
      ],
    });
    mockDispatch.mockReturnValue(jest.fn());

    const { items, count } = useSelector(action.selectCart);
    const { id, title, price, imageUrl, sizes, type } = items[0];

    const component = render(
      <CartItem
        id={id}
        title={title}
        price={price}
        imageUrl={imageUrl}
        count={count}
        sizes={sizes}
        type={type}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  it('Диспат action', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValue(dispatch);

    const mockedClearItemComplete = jest.spyOn(action, 'clearItem');
    const mockedAddItemComplete = jest.spyOn(action, 'addItem');
    const mockedRemoveItemComplete = jest.spyOn(action, 'removeItem');

    const { items, count } = useSelector(action.selectCart);
    const { id, title, price, imageUrl, sizes, type } = items[0];

    render(
      <CartItem
        id={id}
        title={title}
        price={price}
        imageUrl={imageUrl}
        count={count}
        sizes={sizes}
        type={type}
      />,
    );

    fireEvent.click(screen.getByTestId('clear'));

    expect(dispatch).toHaveBeenCalledTimes(1);

    expect(mockedClearItemComplete).toHaveBeenCalledTimes(1);
    expect(mockedClearItemComplete).toHaveBeenCalledWith({ item: { id } });

    fireEvent.click(screen.getByTestId('add'));

    expect(mockedAddItemComplete).toHaveBeenCalledTimes(1);
    expect(mockedAddItemComplete).toHaveBeenCalledWith({
      item: { id, title, price, imageUrl, sizes, type, count },
    });

    fireEvent.click(screen.getByTestId('remove'));

    expect(mockedRemoveItemComplete).toHaveBeenCalledTimes(1);
    expect(mockedRemoveItemComplete).toHaveBeenCalledWith({ item: { id } });
  });
});
