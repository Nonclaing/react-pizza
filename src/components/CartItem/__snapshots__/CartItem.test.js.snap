// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест компонента CartItem Рендер 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart__item"
      >
        <div
          class="cart__item-img"
        >
          <img
            alt="Pizza"
            class="pizza-block__image"
            loading="lazy"
            src="https://media.dodostatic.net/image/r:584x584/11EE7D6130241E75B0AB33725248C0D0.avif"
          />
        </div>
        <div
          class="cart__item-info"
        >
          <h3>
            Пепперони Фреш с перцем
          </h3>
          <p>
            тонкое тесто, 26 см.
          </p>
        </div>
        <div
          class="cart__item-count"
        >
          <div
            class="button button--outline button--circle cart__item-count-minus"
            data-testid="remove"
          >
            <svg
              fill="none"
              height="10"
              viewBox="0 0 10 10"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
                fill="#EB5A1E"
              />
              <path
                d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
                fill="#EB5A1E"
              />
            </svg>
          </div>
          <b>
            1
          </b>
          <div
            class="button button--outline button--circle cart__item-count-plus"
            data-testid="add"
          >
            <svg
              fill="none"
              height="10"
              viewBox="0 0 10 10"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
                fill="#EB5A1E"
              />
              <path
                d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
                fill="#EB5A1E"
              />
            </svg>
          </div>
        </div>
        <div
          class="cart__item-price"
        >
          <b>
            803
             ₽
          </b>
        </div>
        <div
          class="cart__item-remove"
          data-testid="clear"
        >
          <div
            class="button button--outline button--circle"
          >
            <svg
              fill="none"
              height="10"
              viewBox="0 0 10 10"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
                fill="#EB5A1E"
              />
              <path
                d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
                fill="#EB5A1E"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart__item"
    >
      <div
        class="cart__item-img"
      >
        <img
          alt="Pizza"
          class="pizza-block__image"
          loading="lazy"
          src="https://media.dodostatic.net/image/r:584x584/11EE7D6130241E75B0AB33725248C0D0.avif"
        />
      </div>
      <div
        class="cart__item-info"
      >
        <h3>
          Пепперони Фреш с перцем
        </h3>
        <p>
          тонкое тесто, 26 см.
        </p>
      </div>
      <div
        class="cart__item-count"
      >
        <div
          class="button button--outline button--circle cart__item-count-minus"
          data-testid="remove"
        >
          <svg
            fill="none"
            height="10"
            viewBox="0 0 10 10"
            width="10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
              fill="#EB5A1E"
            />
            <path
              d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
              fill="#EB5A1E"
            />
          </svg>
        </div>
        <b>
          1
        </b>
        <div
          class="button button--outline button--circle cart__item-count-plus"
          data-testid="add"
        >
          <svg
            fill="none"
            height="10"
            viewBox="0 0 10 10"
            width="10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
              fill="#EB5A1E"
            />
            <path
              d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
              fill="#EB5A1E"
            />
          </svg>
        </div>
      </div>
      <div
        class="cart__item-price"
      >
        <b>
          803
           ₽
        </b>
      </div>
      <div
        class="cart__item-remove"
        data-testid="clear"
      >
        <div
          class="button button--outline button--circle"
        >
          <svg
            fill="none"
            height="10"
            viewBox="0 0 10 10"
            width="10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.92001 3.84V5.76V8.64C5.92001 9.17016 5.49017 9.6 4.96001 9.6C4.42985 9.6 4.00001 9.17016 4.00001 8.64L4 5.76L4.00001 3.84V0.96C4.00001 0.42984 4.42985 0 4.96001 0C5.49017 0 5.92001 0.42984 5.92001 0.96V3.84Z"
              fill="#EB5A1E"
            />
            <path
              d="M5.75998 5.92001L3.83998 5.92001L0.959977 5.92001C0.429817 5.92001 -2.29533e-05 5.49017 -2.29301e-05 4.96001C-2.2907e-05 4.42985 0.429817 4.00001 0.959977 4.00001L3.83998 4L5.75998 4.00001L8.63998 4.00001C9.17014 4.00001 9.59998 4.42985 9.59998 4.96001C9.59998 5.49017 9.17014 5.92001 8.63998 5.92001L5.75998 5.92001Z"
              fill="#EB5A1E"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
